pkg/identityserver/gormstore/eui_block_store.go

```diff
@@ -78,22 +78,15 @@ func (s *euiStore) issueDevEUIAddressFromBlock(ctx context.Context) (*types.EUI6
 	var euiBlock EUIBlock
 	for {
 		// Get current value from DevEUI db.
-		query := s.query(ctx, EUIBlock{}).Where(EUIBlock{Type: "dev_eui"})
+		query := s.query(ctx, EUIBlock{}).
+			Where(`"eui_blocks"."type"='dev_eui' AND "eui_blocks"."current_counter" <= "eui_blocks"."end_counter"`)
 		if err := query.First(&euiBlock).Error; err != nil {
-			return nil, err
-		}
-		// Check if DevEUI block limit exists.
-		result := s.query(ctx, EUIBlock{}).
-			Where(`"eui_blocks"."type"='dev_eui' AND "eui_blocks"."current_counter" <= "eui_blocks"."end_counter"`).
-			Update("current_counter", gorm.Expr("current_counter + ?", 1))
-		if err := result.Error; err != nil {
-			return nil, err
-		}
-		if result.RowsAffected == 0 {
 			return nil, errMaxGlobalEUILimitReached.New()
 		}
-
-		devEUIResult.UnmarshalNumber(euiBlock.StartEUI.toPB().MarshalNumber() | uint64(euiBlock.CurrentCounter))
+		euiBlock.CurrentCounter++
```

Comment:
This is not atomic. Why is the previous `Update` approach no longer used ?

Reply:
Yeah in current implementation this is still locking the whole table. Once we update to Gorm v2 we could easily use `FOR UPDATE` and just lock the specific row of the table.

Reply:
I reverted this because indeed it seems it could race.